python/python/lance/arrow.py

```diff
@@ -11,14 +11,428 @@
+        <lance.arrow.ImageURIArray object at 0x...>
+        ['file::///tmp/1.png']
+        """
+        from urllib.parse import urlparse
+
+        if not isinstance(uris, pa.StringArray):
+            uris = pa.array((str(uri) for uri in uris), type=pa.string())
+        else:
+            raise TypeError("Cannot build a ImageURIArray from {}".format(type(uris)))
+
+        for uri in uris:
+            # This is to catch malformed URIs
+            parsed_uri = urlparse(uri.as_py())
+            if not parsed_uri.scheme and not parsed_uri.scheme == "":
+                raise ValueError("URI {} is not a valid URI".format(uri))
+
+        return cls.from_storage(ImageURIType(), uris)
+
+    def read_uris(self):
+        """
+        Read the images from the URIs into memory and return an EncodedImageArray
+
+        Returns
+        -------
+        EncodedImageArray
+            Array of encoded images
+
+        Examples
+        --------
+        >>> import os
+        >>> uris = [os.path.join(os.path.dirname(__file__), "../tests/images/1.png")]
+        >>> uri_array = ImageURIArray.from_uris(uris)
+        >>> uri_array.read_uris()
+        <lance.arrow.EncodedImageArray object at 0x...>
+        [...]
+        """
+        from pyarrow import fs
+
+        images = []
+        for uri in self.storage:
+            filesystem, path = fs.FileSystem.from_uri(uri.as_py())
+            with filesystem.open_input_stream(path) as f:
+                images.append(f.read())
+
+        return EncodedImageArray.from_storage(
+            EncodedImageType(), pa.array(images, type=pa.binary())
+        )
+
+
+class EncodedImageArray(ImageArray):
+    """
+    Array of encoded images. Images may be encoded in any format. Format is not stored
+    by the array but can typically be inferred from the image bytes. Alternatively a
+    separate record of encoding can be kept in a separate array outside this library.
+    """
+
+    import numpy as np
+
+    def to_tensor(
+        self, decoder: Optional[Callable[[pa.BinaryArray], np.ndarray]] = None
+    ):
+        """
+        Decode encoded images and return a FixedShapeImageTensorArray
+
+        Parameters
+        ----------
+        decoder : Callable[pa.binary()], optional
+            A function that takes a pa.binary() and returns a numpy.ndarray
+            or pa.fixed_shape_tensor. If not provided, will attempt to use
+            tensorflow and then pillow decoder in that order.
+
+        Returns
+        -------
+        FixedShapeImageTensorArray
+            Array of images as tensors
+
+        Examples
+        --------
+        >>> import os
+        >>> uris = [os.path.join(os.path.dirname(__file__), "../tests/images/1.png")]
+        >>> encoded_image_array = ImageURIArray.from_uris(uris).read_uris()
+        >>> encoded_image_array.to_tensor()
+        <lance.arrow.FixedShapeImageTensorArray object at 0x...>
+        [[42, 42, 42, 255]]
+        """
+        import numpy as np
```

Comment:
Small UX, before you decode the images, it would be nice to tell them up front if their PyArrow is too old:


```suggestion
        import numpy as np
        
        if not hasattr(pa, "FixedShapeTensorType"):
            raise NotImplementedError("This function requires PyArrow >= 12.0.0")
```